feat(metrics): simplify custom application metrics (#2716)
SumoLogic/sumologic-kubernetes-collection: docs/collecting-application-metrics.md

## Practical scenarios

### Application metrics are exposed (one endpoint scenario)

If there is only one endpoint in the pod you want to scrape metrics from, you can use annotations.
Add the following annotations to your pod definition:

```yaml
# ...
annotations:
  prometheus.io/port: 8000 # Port which metrics should be scraped from
  prometheus.io/scrape: true # Set if metrics should be scraped from this pod
  prometheus.io/path: "/metrics" # Path which metrics should be scraped from
```

**NOTE:** If you add more than one annotation with the same name, only the last one will be used.

Next, add the following configuration to your `user-values.yaml`:

```yaml
kube-prometheus-stack:
  prometheus:
    prometheusSpec:
      additionalRemoteWrite:
        - url: http://$(METADATA_METRICS_SVC).$(NAMESPACE).svc.cluster.local.:9888/prometheus.metrics.<custom endpoint name>
          writeRelabelConfigs:
          - action: keep
            regex: <metric1>|<metric2>|...
            sourceLabels: [__name__]
```

**Note:** We recommend to use regex validator, for example [https://regex101.com/].

### Application metrics are exposed (multiple enpoints scenario)

> If in the pod there is more than one endpoint you want to scrape metrics from,

Reviewer: ```suggestion
If you want to scrape metrics from multiple endpoints in a single Pod,
```

in your `user-values.yaml`:

```yaml
kube-prometheus-stack:
  prometheus:
    additionalServiceMonitors:
      - name: <service monitor name>
        endpoints:
          - port: <port name or number>
            path: <metrics path>
        namespaceSelector:
          matchNames:
            - <namespace>
        selector:
          matchLabels:
            <identyfing label 1>: <value of indentyfing label 1>
            <label2>: <value of identyfing label 2>
```

**Note** For advanced serviceMonitor configuration, please look at the [Prometheus documentation][prometheus_service_monitors]

At the end, you need to add the following configuration to your `user-values.yaml`,
to instruct Prometheus which metrics should be forwarded through the pipeline.

```yaml
kube-prometheus-stack:
  prometheus:
    prometheusSpec:
      additionalRemoteWrite:
        - url: http://$(METADATA_METRICS_SVC).$(NAMESPACE).svc.cluster.local.:9888/prometheus.metrics.<custom endpoint name>
          writeRelabelConfigs:
          - action: keep
            regex: <metric1>|<metric2>|...
            sourceLabels: [__name__]
```

> **Note:** We recommend to use regex validator, for example [https://regex101.com/]

Reviewer: ```suggestion
**Note:** We recommend using a regex validator, for example [https://regex101.com/]
```

# Collecting application metrics

> This document is going to cover multiple different use cases related to scraping custom application metrics.

Reviewer 2: ```suggestion
This document covers multiple different use cases related to scraping custom application metrics.
```

# Collecting application metrics

> This document covers multiple different use cases related to scraping custom application metrics.

Reviewer 2: ```suggestion
This document covers multiple different use cases related to scraping custom application metrics exposed in Prometheus format.
```

Author: It's not exactly true, but I will make few more iterations on that doc, so we can improve it later
